flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/ForStKeyedStateBackend.java

```diff
@@ -445,6 +445,22 @@ private <N, UK, SV> RegisteredKeyValueStateBackendMetaInfo<N, SV> updateRestored
 
         RegisteredKeyValueStateBackendMetaInfo<N, SV> restoredKvStateMetaInfo = oldStateInfo.f1;
 
+        // fetch current namespace serializer now because if it is incompatible, we can't access
+        // it anymore to improve the error message
+        TypeSerializer<N> previousNamespaceSerializer =
+                restoredKvStateMetaInfo.getNamespaceSerializer();
+
+        TypeSerializerSchemaCompatibility<N> s =
+                restoredKvStateMetaInfo.updateNamespaceSerializer(namespaceSerializer);
+        if (s.isCompatibleAfterMigration() || s.isIncompatible()) {
```

Comment:
I am curious about the difference between` isCompatibleAfterMigration()` and `isIncompatible()`. If the difference is meaningful, should we include these different incompatibilities in the error message to help make the message more useful?

Reply:
@mayuehappy We could distinguish the messages for two cases. For the `isCompatibleAfterMigration`, we may say 'The new namespace serializer () can be compatible with the old one () after migration. But ForSt does not support state migration yet.'

Reply:
@Zakelly Thanks for the reply. As I remember, `NamespaceSerializer` should not support migration, even RocksdbStateBackend does. So do we still need to distinguish here?

Reply:
@mayuehappy we can keep it if so.